session/pingpong/consumer_balance_tracker.go

```diff
@@ -129,8 +125,13 @@ func (cbt *ConsumerBalanceTracker) handleUnlockEvent(id string) {
 	cbt.ForceBalanceUpdate(identity)
 }
 
-func (cbt *ConsumerBalanceTracker) handleGrandTotalRecovered(ev GrandTotalRecovered) {
-	cbt.ForceBalanceUpdate(ev.Identity)
+func (cbt *ConsumerBalanceTracker) handleGrandTotalChanged(ev AppEventGrandTotalChanged) {
+	if _, ok := cbt.balances[ev.ConsumerID]; !ok {
```

Comment:
Missing lock?

Reply:
actually - no, both of the cases are internally locked aftewards, so we're fine here.